tardis/utilities/executors/sshexecutor.py

```diff
@@ -3,43 +3,107 @@
 from ...interfaces.executor import Executor
 from ..attributedict import AttributeDict
 
+try:
+    from contextlib import asynccontextmanager
+except ImportError:
+    from aiotools import async_ctx_manager as asynccontextmanager
+
 import asyncssh
+import asyncio
+
+
+class SSHConnectionPool(object):
+    _connection_pool = dict()
+    _lock = None
+
+    def __init__(self, connection_pool_size=5, **parameters):
+        assert (
+            "host" in parameters
+        ), f"Host to connect to not found in parameters {parameters}"
+
+        self._connection_pool_size = connection_pool_size
+        self._host = parameters["host"]
+        self._parameters = parameters
+
+    @property
+    def acquire_lock(self):
+        # Create lock once tardis event loop is running.
+        # To avoid got Future <Future pending> attached to a different loop exception
+        if not self._lock:
+            self._lock = asyncio.Lock()
+        return self._lock
+
+    async def establish_connection(self):
+        for retry in range(1, 10):
+            try:
+                return await asyncssh.connect(**self._parameters)
+            except (
+                ConnectionResetError,
+                asyncssh.DisconnectError,
+                asyncssh.ConnectionLost,
+                BrokenPipeError,
+            ):
+                await asyncio.sleep(retry * 10)
+        return await asyncssh.connect(**self._parameters)
+
+    async def _init_connection_pool(self):
+        async with self.acquire_lock:
+            # check that connection pool has not yet been initialize in a different task
+            if self._host not in self._connection_pool:
+                self._connection_pool[self._host] = asyncio.Queue()
+                while (
+                    self._connection_pool[self._host].qsize()
+                    < self._connection_pool_size
+                ):
+                    await self._connection_pool[self._host].put(
+                        await self.establish_connection()
+                    )
```

Comment:
This seems vulnerable to a race condition. As soon as ``self._connection_pool[self._host]`` is assigned, other tasks will skip initialisation (L64) and try to ``.get()`` connections. If there are N tasks, then worst-case 1 does the initialisation and N-1 get connections as soon as they are available. The pool may thus be up to N-1 elements smaller than created connections (since they are temporarily owned by the tasks).

Once initialisation is done, the pool contains connection_pool_size connections, and may then get the additional N-1 connections returned, for a total of connection_pool_size + N - 1 connections.

Directly creating the appropriate number seems more robust:
```suggestion
                self._connection_pool[self._host] = asyncio.Queue()
                for _ in range(self._connection_pool_size):
                    await self._connection_pool[self._host].put(
                        await self.establish_connection()
                    )
```